### PR TITLE
[PF-1753] Add highest role to workspace returned by workspace actions

### PR DESCRIPTION
**Pull request by melissachang**
ListWorkspaces works in swagger:

image

Also tested in swagger: GetWorkspace, GetWorkspaceByUserFacingId, UpdateWorkspace

FYI there was one test where workspace had APPLICATION role: [PrivateResourceCleanupServiceTest. cleanupResourcesSuppressExceptions_cleansApplicationPrivateResource_succeeds()](https://github.com/DataBiosphere/terra-workspace-manager/blob/ff8d0ac9e043ff58e97afb68faf5a1836d1c7685/service/src/test/java/bio/terra/workspace/service/privateresource/PrivateResourceCleanupServiceTest.java#L239)

### CODE REVIEW

File: service/src/main/java/bio/terra/workspace/service/iam/model/WsmIamRole.java [hunk 1]
```diff
@@ -38,6 +42,32 @@ public static WsmIamRole fromSam(String samRole) {
...
   }
 
+  public static WsmIamRole getHighestRole(UUID workspaceId, List<WsmIamRole> roles) {
+    if (roles.isEmpty()) {
+      throw new InternalServerErrorException(
+          String.format("Workspace %s missing roles", workspaceId.toString()));
+    }
+
+    if (roles.contains(WsmIamRole.APPLICATION)) {
+      Preconditions.checkState(
```
Comment (ddietterich) [inline service/src/main/java/bio/terra/workspace/service/iam/model/WsmIamRole.java:52]: Why do we need to require that an application only have APPLICATION role?
I think you get the same effect if you decide that APPLICATION collates highest.

File: openapi/src/parts/workspace.yaml [hunk 2]
```diff
@@ -476,6 +476,11 @@ components:
         description:
           description: A description of the workspace
           type: string
+        highestRole:
```
Comment (cbookg) [inline openapi/src/parts/workspace.yaml:479]: should this be required?

**Review by ddietterich — CHANGES_REQUESTED** [review]
I do not think the highestRole should be in the workspace object. Ideally, that object is not mutated after being fetched from the DB. Instead, I think highest role should only be fetched from the controller when we need to return the data to the caller. We never use it internally to WSM.

File: service/src/main/java/bio/terra/workspace/service/iam/SamService.java [hunk 3]
```diff
@@ -371,6 +377,7 @@ public boolean checkAuthAsWsmSa(
    * @param resourceType The Sam type of the resource being checked
    * @param resourceId The ID of the resource being checked
    * @param action The action being checked on the resource
+   * @return
```
[the remark this answers is not in the document]
Comment (melissachang) [reply service/src/main/java/bio/terra/workspace/service/iam/SamService.java]: Done

**Review by ddietterich — APPROVED** [review]
LGTM, thank you!

File: service/src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java [hunk 4]
```diff
@@ -139,11 +139,27 @@ public Workspace validateWorkspaceAndAction(
...
+  @Traced
+  private Workspace validateWorkspaceAndAction_inner(
+      AuthenticatedUserRequest userRequest, Workspace workspace, String action) {
+    UUID workspaceUuid = workspace.getWorkspaceId();
     SamRethrow.onInterrupted(
         () ->
             samService.checkAuthz(
                 userRequest, SamConstants.SamResource.WORKSPACE, workspaceUuid.toString(), action),
         "checkAuthz");
+
```
[the remark this answers is not in the document]
Comment (ddietterich) [reply service/src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java]: I wish the layering was cleaner between service and controller. We end up with more controller code than I would like; especially in the resource methods. The benefit is that the resource objects are immutable.

**Review by yuhuyoyo — APPROVED** [review]

File: openapi/src/parts/workspace.yaml [hunk 5]
```diff
@@ -476,6 +476,11 @@ components:
...
+          description: |
+            Highest role. If requester is both reader and writer, writer will be
```
[the remark this answers is not in the document]
Comment (melissachang) [reply openapi/src/parts/workspace.yaml]: Done

File: service/src/main/java/bio/terra/workspace/service/iam/model/WsmIamRole.java [hunk 6]
```diff
@@ -38,6 +41,25 @@ public static WsmIamRole fromSam(String samRole) {
...
 
+  public static WsmIamRole getHighestRole(UUID workspaceId, List<WsmIamRole> roles) {
+    if (roles.isEmpty()) {
+      throw new InternalServerErrorException(
+          String.format("Workspace %s missing roles", workspaceId.toString()));
+    }
+
+    if (roles.contains(WsmIamRole.APPLICATION)) {
+      return WsmIamRole.APPLICATION;
+    } else if (roles.contains(WsmIamRole.OWNER)) {
```
[the remark this answers is not in the document]
Comment (melissachang) [reply service/src/main/java/bio/terra/workspace/service/iam/model/WsmIamRole.java:52]: `else if` after `return` looks pretty common, so I'll leave as is -- It's a little more clear to me (though you're right that the `else` doesn't do anything).

https://cs.github.com/DataBiosphere/terra-workspace-manager?q=return+%22else+if%22

**Pull request merged**